docs/posts/duckdb-for-rag/index.qmd

```diff
@@ -0,0 +1,543 @@
+> always try to provide models with the appropriate context to provide high
+> quality answers." - [Hamel
+> Husain](https://twitter.com/HamelHusain/status/1709740984643596768)
+
+RAG has become synonymous with turning your text data into "embeddings"
+(numbers), running a similarity search (e.g. cosine similarity) over those
+numbers against input text, and taking the top `k` results to feed into the LLM.
+This loses out on the semantic understanding LLMs are so powerful for and often
+results in bespoke technology stacks that are unnecessary.
+
+### Context length issues
+
+When GPT-4 launched, it supported a context length of about 8,000 tokens.
+Typically, one word is about two tokens, so we can call this about 4,000 input
+words. While that is a lot of words, much of the length gets taken up by the
+**system prompt** that instructs the LLM on what to do and other metadata that
+may be used as input. In practice, 8k token context length is not much.
+
+Over time, these context lengths have been increasing -- 16k, then 32k, and now
+128k for GPT-4-turbo models. Other LLMs have even higher context lengths, with
+Google's Gemini family reporting 1M and 10M token context lengths. However, even
+these large context lengths do not solve some key problems. For one, LLMs may
+ignore or even be confused by context. While you can feed in a full textbook on
+machine learning, if you're just asking for information on k-means clustering
+you should probably only include that chapter. Additionally, you are charged
+money by the number of input tokens (and response tokens) and the time it takes
+to inference is proportional to that number. Regardless of the context length of
+the model, for the foreseeable future you want to be as efficient as possible.
+
+Thus, you need to have some method of retrieving and augmentign your input text
+with the most relevant information for the task at hand.
+
+## Setup
+
+Another reason I dislike RAG is, like traditional ML (and even software
+engineering), is that the best approach is often "**it depends**". What are you
+trying to accomplish? What are your constraints on resources?
+
+### Scenario
+
+For this blog, we'll use a scenario: we are trying to build a state-of-the-art
+exploratory data analysis (EDA) chatbot. I want to be able to ask it questions
+about my data, have it generate accurate SQL, and run that SQL on my behalf.
+
+We'll be using toy data (our favorite penguins dataset!) and LLM-generated
+English and SQL queries, but the same principles should apply for real-world use
+cases.
+
+### Setup code
+
+First, we need to install the necessary packages:
+
+```bash
+pip install 'ibis-framework[duckdb,examples]' openai marvin python-dotenv
+```
+
+Then import them:
+
+```{python}
+import ibis
+import marvin
+
+from openai import OpenAI
+
+from dotenv import load_dotenv
+```
+
+You need to set the `OPENAI_API_KEY` environment variable to your OpenAI API
```

Comment:
In the install section, we are talking in the "We" person and here you say "You" need, you want to pick one and be consistent. Or remove the person entirely.